python/src/trezorlib/authentication.py

```diff
@@ -0,0 +1,303 @@
+            if tag != b"\x02":
+                raise ValueError("Invalid DER signature: not an integer.")
+            length = reader.read(1)[0]
+            if length > 33:
+                raise ValueError("Invalid DER signature: integer too long.")
+            return int.from_bytes(reader.read(length), "big")
+
+        r = read_int()
+        s = read_int()
+        if reader.tell() != len(sig_bytes):
+            raise ValueError("Invalid DER signature: trailing data.")
+        return r, s
+
+    @staticmethod
+    def fix_signature(sig_bytes: bytes) -> bytes:
+        r, s = Certificate._decode_signature_permissive(sig_bytes)
+        reencoded = utils.encode_dss_signature(r, s)
+        if reencoded != sig_bytes:
+            LOG.info(
+                "Re-encoding malformed signature: %s -> %s",
+                sig_bytes.hex(),
+                reencoded.hex(),
+            )
+        return reencoded
+
+
+def verify_authentication_response(
+    challenge: bytes,
+    signature: bytes,
+    cert_chain: t.Iterable[bytes],
+    *,
+    whitelist: t.Collection[bytes] | None,
+    allow_development_devices: bool = False,
+    root_pubkey: bytes | ec.EllipticCurvePublicKey | None = None,
+) -> None:
+    """Evaluate the response to an AuthenticateDevice call.
+
+    Performs all steps and logs their results via the logging facility. (The log can be
+    accessed via the `LOG` object in this module.)
+
+    When done, raises DeviceNotAuthentic if the device is not authentic.
+
+    The optional argument `root_pubkey` allows you to specify a root public key either
+    as an `ec.EllipticCurvePublicKey` object or as a byte-string representing P-256
+    public key.
+    """
+    if isinstance(root_pubkey, (bytes, bytearray, memoryview)):
+        root_pubkey = ec.EllipticCurvePublicKey.from_encoded_point(
+            ec.SECP256R1(), root_pubkey
+        )
+
+    challenge_bytes = (
+        len(CHALLENGE_HEADER).to_bytes(1, "big")
+        + CHALLENGE_HEADER
+        + len(challenge).to_bytes(1, "big")
+        + challenge
+    )
+
+    cert_chain_iter = iter(cert_chain)
+
+    failed = False
+
+    try:
+        cert = Certificate(next(cert_chain_iter))
+    except Exception:
+        LOG.error("Failed to parse device certificate.")
+        raise DeviceNotAuthentic
+
+    try:
+        cert.verify(signature, challenge_bytes)
+    except exceptions.InvalidSignature:
+        LOG.error("Challenge verification failed.")
+        failed = True
+    else:
+        LOG.debug("Challenge verified successfully.")
+
+    cert_label = "Device certificate"
+    for i, issuer_bytes in enumerate(cert_chain_iter, 1):
+        try:
+            ca_cert = Certificate(issuer_bytes)
+        except Exception:
+            LOG.error(f"Failed to parse CA certificate #{i}.")
+            failed = True
+            continue
+
+        if whitelist is None:
+            LOG.warning("Skipping public key whitelist check.")
+        else:
```

Comment:
What is the usecase for `whitelist`?

Reply:
linked to https://github.com/trezor/data/pull/79
basically you want to download a whitelist of public keys and check that your Trezor is included. this allows us to "revoke" a batch of Trezors for example